### ISSUE TITLE
-exclude flag for targeted plan/apply

### ISSUE THREAD

**Issue by JamesBalazs** [opentofu/opentofu#426]
### Summary

This RFC details a new flag, `-exclude`, which would do the inverse of `-target` - that is, plan/apply all resources _except_ those specified by the exclude flag and their descendants.

### Problem Statement

Only for use in exceptional circumstances as is the case for `-target`, this makes it a lot easier to plan/apply all changes to an environment except ones with known issues eg where there is a problem with the infrastructure or the committed config for one module/resource, but you still want to apply everything else.

### User-facing description

Usage example:
`opentf plan -exclude module.something_broken` / `opentf apply -exclude module.something_broken`

The `-exclude` feature could be explained as an argument that allows you to plan/apply all resources except a specified resource.

### Technical Description

- `-exclude` flag added to `extendedFlagSet` when an operation is provided (same as `-target`)
- Passed through to `OperationRequest`
- `TargetsTransformer.Transform` should remove vertices that _are_ in `t.Excludes` if `t.Excludes` are provided (just like how it removes vertices that _aren't_ in `t.Targets`).
- If both targets and excludes are provided, use the most specific path eg:
  ``` 
  -exclude module.entire_environment \
  -include module.entire_environment.module.network \
  -include module.entire_environment.module.application \
  -exclude module.entire_environment.module.application.module.eks
  ```
  should exclude all of the entire_environment module but include the network and application modules within, _except_ the eks module within the application module. This should be the outcome regardless of the order of the flags.
- Return an error when -target and -exclude point at the exact same resource
- When destroying, if there are dependencies of an excluded resource, the dependencies should _also_ be kept (excluded from change)

### Rationale and alternatives

The rationale is that this is a highly requested tf feature, which makes certain scenarios (eg a single resource in a module is broken, but you still need to apply the module) potentially easier to resolve since you can exclude the offending resource.

### Downsides

I can't think of any downsides, unless opentf is intended to support the exact same flags as tf.

### Unresolved Questions

In my poc branch I've added the flag, and added the `TargetsTransformer` change that should make excludes work. I've also followed through a few places where targets are passed around and added excludes there too.

However I'm not too familiar with the opentf architecture other than a brief look at the doc, so there may be things I've missed to make this work, and places I've added excludes that didn't really need it. Any pointers from people familiar with the code would be appreciated.

Also since we can't change TFE which is imported from a Hashicorp project, I presume there would be backends that we can't support this flag for? At the very least I guess this would require me to add errors when somebody tries to pass `-exclude` using unsupported backends? If not a complete blocker for this feature ever going in.

### Related Issues

https://github.com/hashicorp/terraform/issues/2253 this issue in the original Terraform repo requests an exclude flag that does the inverse of `-target`, and is very highly requested.

### Proof of Concept

https://github.com/opentffoundation/opentf/compare/main...JamesBalazs:opentf-excludes:add-exclude-flag is a fork (POC PR https://github.com/opentffoundation/opentf/pull/427) with what I _think_ is the necessary (surprisingly simple) change to `TargetsTransformer.Transform`, and the flag added to `extendedFlagSet`. I have not been able to test if this works yet since I don't have anything using tf on my home macbook, and I threw this together at 2am so it still needs unit tests etc adding.

**Comment by Hefeweizen** [comment]
> If both targets and excludes are provided it can either:
>  -  Have excludes override targets (plan/apply this list of targets, except this list of excludes)
>  -  Return an error (-targets and -excludes can't be used at the same time)
>    The former is easier to implement, since it looks like it should just work, and more flexible (although I can't imagine a use-case)

I can imagine the use case of "-exclude module that has eks, s3 buckets for logs from eks compoents; -target s3 buckets".  Perhaps in standing up new infrastructure, the s3 buckets need to be in place but the operator doesn't want to wait on the 15+ minutes for the entire eks cluster to build.

At some level, this kind of action is going against the tool.  The tool builds its own DAG, and wants to figure out the ordering.  If a user is consistently doing this, they probably need to move the creation of the s3 buckets into the appropriate layer.

While I can appreciate these arguments, I don't support it.  The user using the tool has better context of what they need, and a better understanding of their desire.  I think we have smart users and we shouldn't complicate their lives by limiting what they can do.  After all, why have -exclude in the first place if we're not trying to make their lives easier?

As for behaviour if both `-target` and `-exclude` is specified, I'd suggest we take ordering from the command line invocation.  This would allow very precise targetting.  For example:
```
opentf plan \
  -exclude module.entire_environment \
  -include module.entire_environment.module.network \
  -include module.entire_environment.module.application \
  -exclude module.entire_environment.module.application.module.eks
```


> Also since we can't change TFE which is imported from a Hashicorp project, I presume there would be backends that we can't support this flag for? At the very least I guess this would require me to add errors when somebody tries to pass -exclude using unsupported backends? If not a complete blocker for this feature ever going in.

Could you elaborate on your thoughts here?  I'm not sure I understand the concern, as from my reading, `TargetsTransformer.Transform` happens downstream of the backends doing their work.

**Comment by Hefeweizen** [comment]
> Also since we can't change TFE which is imported from a Hashicorp project, I presume there would be backends that we can't support this flag for? At the very least I guess this would require me to add errors when somebody tries to pass -exclude using unsupported backends? If not a complete blocker for this feature ever going in.

If I'm reading the code right, it seems backends do try and process targets, but for local and remote, it's mostly to say if the statefile version supports targetting (and why it really cares, I can't say).

`internal/cloud/backend_plan.go` and `internal/cloud/backend_plan.go` both have lines like `runOptions.TargetAddrs = append(runOptions.TargetAddrs, addr.String())` where they maintain the cli options, but I can't see where they actually do anything with this.

**Comment by JamesBalazs** [comment]
> While I can appreciate these arguments, I don't support it. The user using the tool has better context of what they need, and a better understanding of their desire. I think we have smart users and we shouldn't complicate their lives by limiting what they can do. After all, why have -exclude in the first place if we're not trying to make their lives easier?

I agree that a lot of use cases for this can probably be resolved by "just" moving resources to the correct layer, using `depends_on` where appropriate etc.

But we're also both on the same page that users shouldn't be restricted in what they can do with the tool simply because it isn't always best practice to exclude resources from an apply. It might not be best practice to target resources either, yet it's often incredibly useful to do so - I guess the exclude flag is mostly about making tf more flexible even if it's only the right solution in some edge cases.

**Comment by cube2222** [comment]
Thanks for submitting the issue @JamesBalazs!

There are people doing this today by using a ton of -target flags, so I think this is very useful for companies onboarding Tofu from scratch (with part of the infra not working).

Overall, we're keen to accept this, we just need to review it in detail. One thing we're worried about is what happens in the scenario of destroy with exclude if there are dependencies of the excluded resource. They should probably be kept.

**Comment by cube2222** [comment]
One more edge-case to consider, what if somebody specifies **both** exclude and target?

I think that makes sense, and we should only process targeted resources that aren't excluded.

**Comment by JamesBalazs** [comment]
@cube2222 it looks like @Hefeweizen had a suggestion that if somebody specifies both exclude and target, we should:

> take ordering from the command line invocation. This would allow very precise targetting

Correct me if I'm wrong, but I don't think ordering should matter here. If you have
``` 
  -exclude module.entire_environment \
  -target module.entire_environment.module.network
```
then I would expect we should prioritise the _most specific_ path, rather than prioritising based on order.

@janosdebugs I've updated the description of the RFC to reflect this and to include @cube2222's concern about keeping dependencies when excluding the parent resource.

**Comment by RLRabinowitz** [comment]
Hey @JamesBalazs, I'd like to lay out how to `-exclude` flag should behave, in order to have a specific design for the `-exclude` flag in this issue:

- The `-exclude` flag should mark resources and modules that should not take part in the plan process
- During plan, the plan should not include "excluded" resources, and all **resources dependent on them** (The descendants in the graph). That is the inverted behaviour when compared to the `-target` flag, which includes the "targeted" resource and all resources that are dependencies of it
- No need to support the `-exclude` flag in remote backend plan/apply executions from local (when remotely invoking plans/applies in remote backend / cloud), although the `-target` flag is supported. This is because that the `go-tfe` library defines that API used with the remote backend / cloud, and we do not want to fork it and change the API at this time
- If the `-exclude` flag gets a reference value that does not exist in the plan, then it should technically just be ignored. So, a plan with a single `-exclude` flag with a non-existent reference would result in a full plan
- Like the `-target` flag, using the `-exclude` flag should produce a warning on plan action
- When using the `-exclude` flag with the `-target` flag, the plan should include all targeted resources and all resources that are their dependecies (ancestors in the graph), but from these - remove all excluded resources and all resources dependent on them (descendants in the graph)

**Comment by JamesBalazs** [comment]
> Thank you for your patience @JamesBalazs . We have accepted the issue. We have recently changed our RFC process to be PR driven, creating an `.md` file that will exist in the repository, which allows having discussions on each decision in the RFC process in an easier way.
> 
> Would you mind following the new [contribution guidelines on how to create an RFC](https://github.com/opentofu/opentofu/blob/main/rfc/README.md) to backfill this information? I'm also willing to do that for you if you'd like, just thought you'll want to be involved in that 🙃

I created an RFC via the new process, and copied as much information over from this issue as possible. Please let me know if there is any extra detail or changes needed.

**Comment by 16g** [comment]
Perfect!
 🎉

**Comment by cam72cam** [comment]
Tentatively added as a stretch goal in the 1.9 milestone.

**Comment by RLRabinowitz** [comment]
Thank you for your contributions with helping flesh out this issue and the RFC @JamesBalazs !

I've created an RFC for the `-exclude` flag in https://github.com/opentofu/opentofu/pull/1860

**Comment by pdfrod** [comment]
This was marked as completed, but it seems it was just the RFC? Where can I follow the actual implementation?

**Comment by cam72cam** [comment]
Looks like it auto closed due to the incorrect "Resolved: " tag, thanks for catching that!
